include/onnxruntime/core/framework/float16.h

```diff
@@ -17,53 +27,64 @@ struct MLFloat16 {
 
   float ToFloat() const;
 
-  operator float() const {
-    return ToFloat();
-  }
+  operator float() const { return ToFloat(); }
 };
 
-inline bool operator==(const MLFloat16& left, const MLFloat16& right) {
-  return left.val == right.val;
-}
+inline bool operator==(const MLFloat16& left, const MLFloat16& right) { return left.val == right.val; }
+inline bool operator!=(const MLFloat16& left, const MLFloat16& right) { return left.val != right.val; }
+inline bool operator<(const MLFloat16& left, const MLFloat16& right) { return left.val < right.val; }
 
-inline bool operator!=(const MLFloat16& left, const MLFloat16& right) {
-  return left.val != right.val;
-}
-
-inline bool operator<(const MLFloat16& left, const MLFloat16& right) {
-  return left.val < right.val;
-}
-
-//BFloat16
+// BFloat16
 struct BFloat16 {
   uint16_t val{0};
-  explicit BFloat16() = default;
-  explicit BFloat16(uint16_t v) : val(v) {}
-  explicit BFloat16(float v) {
+#if defined(USE_ROCM)
+  ORT_HOST_DEVICE BFloat16() = default;
+#else
+  BFloat16() = default;
+#endif
+
+  struct FromBitsT {};
```

Comment:
What is the reason to introduce struct FromBitsT?

Reply:
This idea is from PyTorch. It means if it's initialized from FromBitsT, then the bits will assign to val directly (the real value of BFloat16 instance is not equal to bits), but if not, for example, BFloat16(unsigned short value), it will initialize a BFloat16 == value (but the val member in the object is not equal to value). This is critical for some casting case, for example, BFloat16(1), which casts int to BFloat16, if we don't have this FromBitsT, the complier will report error saying ambiguous constructors, it doesn't know which to choose from BFloat16(unsigned short) or BFloat16(float). Even we don't have such ambigous problem, if compiler chooses BFloat(unsigned short) to do the job but assign the 1 to val memer directly, we would get a wrong BFloat16 instance. Actually our MLFloat16 also has such bug, but we don't have code such as MLFloat16(1) so we haven't encountered the compiler error for now.